KAFKA-16649: Remove lock from DynamicBrokerConfig.removeReconfigurable

cmccabe opened the pull request:
Do not acquire the DynamicBrokerConfig lock in DynamicBrokerConfig.removeReconfigurable. It's not necessary, because the list that these functions are modifying is a thread-safe CopyOnWriteArrayList.  In DynamicBrokerConfig.reloadUpdatedFilesWithoutConfigChange, I changed the code to use a simple Java forEach rather than a Scala conversion, in order to feel more confident that concurrent modifications to the List would not have any bad effects here. (forEach is always safe on CopyOnWriteArrayList.)

cmccabe committed:
KAFKA-16649: Remove lock from DynamicBrokerConfig.removeReconfigurable

Do not acquire the DynamicBrokerConfig lock in DynamicBrokerConfig.removeReconfigurable. It's not
necessary, because the list that these functions are modifying is a thread-safe
CopyOnWriteArrayList.  In DynamicBrokerConfig.reloadUpdatedFilesWithoutConfigChange, I changed the
code to use a simple Java forEach rather than a Scala conversion, in order to feel more confident
that concurrent modifications to the List would not have any bad effects here. (forEach is always
safe on CopyOnWriteArrayList.)

chia7712 commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -303,17 +303,17 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
 
-  def addReconfigurable(reconfigurable: Reconfigurable): Unit = CoreUtils.inWriteLock(lock) {
+  def addReconfigurable(reconfigurable: Reconfigurable): Unit = {
     verifyReconfigurableConfigs(reconfigurable.reconfigurableConfigs.asScala)
```
It seems `verifyReconfigurableConfigs` does not need lock also, since it just check the input strings.

cmccabe replied:
Good catch.

chia7712 replied:
@cmccabe Can we remove the lock from `verifyReconfigurableConfigs` in this PR? Or we can address that in next one.

cmccabe replied:
done

chia7712 commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -1541,6 +1541,36 @@ class KRaftClusterTest {
+              new ConfigEntry(KafkaConfig.NumNetworkThreadsProp, "2"), OpType.SET)))).all().get()
+        val newTopic = Collections.singletonList(new NewTopic("test-topic", 1, 1.toShort))
+        val createTopicResult = admin.createTopics(newTopic)
+        createTopicResult.all().get()
```
It seems reducing the network threads causes the disconnection and it produces the retry of request forward. Hence, that can make this test unstable. Maybe we can increase the network threads instead?

chia7712 commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -1541,6 +1541,36 @@ class KRaftClusterTest {
   }
+
+  @Test
+  def testReduceNumNetworkThreads(): Unit = {
```
Pardon me, is this test case related to deadlock?

cmccabe replied:
I wanted to ensure that no deadlocks are created in the future, by adding this test coverage. I don't think this PR is required to avoid deadlock currently, however.

chia7712 reviewed (commented):
@cmccabe nice improvement. some trivial comments left.

mumrah reviewed (commented):
Thanks for the patch, @cmccabe. I looked for other usages of this lock and see we're obtaining the write lock in `DynamicBrokerConfig#updateBrokerConfig` which gets called from DynamicConfigPublisher. Will we still need this locking when we have dropped ZK? 

Can you add a comment next to the `lock` indicating what it is guarding?

cmccabe commented:
> Thanks for the patch, @cmccabe. I looked for other usages of this lock and see we're obtaining the write lock in DynamicBrokerConfig#updateBrokerConfig which gets called from DynamicConfigPublisher. Will we still need this locking when we have dropped ZK?

I think we can simplify this code a lot once ZK is gone, yes. Having a single updater will help a lot. I suspect we can get rid of the lock at that point, although I'll have to look more later.

> Can you add a comment next to the lock indicating what it is guarding?

I added a clarification that the lock protects against concurrent reconfiguration operations, but not concurrent additions or removals of reconfigurables. As you mentioned, this is mostly only relevant to ZK mode since there we can have racing threads.

mumrah commented:
The new test seems to be failing: [`testReduceNumNetworkThreads() – kafka.server.KRaftClusterTest`](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15838/1/tests) can you check on it?

chia7712 commented:
> The new test seems to be failing: [testReduceNumNetworkThreads() – kafka.server.KRaftClusterTest](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15838/1/tests) can you check on it?

see https://github.com/apache/kafka/pull/15838#discussion_r1585911858

chia7712 commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -1579,6 +1579,36 @@ class KRaftClusterTest {
   }
+
+  @Test
+  def testReduceNumNetworkThreads(): Unit = {
```
Could you please rename it to `testIncreaseNumNetworkThreads` by virtue of recent change?

chia7712 reviewed (approved):
@cmccabe thanks for updated PR. please take a look at one small comment

chia7712 commented:
@cmccabe could you please rebase code to include the fix #16857?

cmccabe commented:
> @cmccabe could you please rebase code to include the fix https://github.com/apache/kafka/pull/16857?

rebased
